File: src/subtypes.test.ts
import { test } from "node:test";

import { expect } from "expect";

import { loadJdkStub } from "./jdkStub.ts";
import { createProgram } from "./program.ts";
import { getSubtypeIndex } from "./subtypes.ts";
import { type Uri } from "./workspace.ts";

test("direct and transitive subtypes resolve across files; memo invalidates on change", () => {
  const program = createProgram();
  loadJdkStub(program);
  program.addProjectFile("file:///I.java" as Uri, "interface I {}");
  program.addProjectFile("file:///A.java" as Uri, "class A implements I {}");
  program.addProjectFile("file:///B.java" as Uri, "class B extends A {}");

  const i = program.getGlobalIndex().getType("I")!;
  const a = program.getGlobalIndex().getType("A")!;
  let index = getSubtypeIndex(program);
  expect(index.directSubtypesOf(i).map(s => s.escapedName)).toEqual(["A"]);
  expect(
    index
      .allSubtypesOf(i)
      .map(s => s.escapedName)
      .sort(),
  ).toEqual(["A", "B"]);
  expect(index.directSubtypesOf(a).map(s => s.escapedName)).toEqual(["B"]);

  // The memo is generation-keyed: a new subtype shows up after a file change.
  program.addProjectFile("file:///C.java" as Uri, "class C implements I {}");
  index = getSubtypeIndex(program);
  expect(index.allSubtypesOf(program.getGlobalIndex().getType("I")!).length).toBe(3);
});

File: togo/internal/services/subtypes_test.go
package services

import (
	"sort"
	"testing"

	"github.com/nikeee/cappu/internal/compiler"
)

// Port of src/services/subtypes.test.ts.

func names(syms []*compiler.Symbol) []string {
	out := []string{}
	for _, s := range syms {
		out = append(out, s.EscapedName)
	}
	return out
}

func TestSubtypeIndexAcrossFiles(t *testing.T) {
	program := compiler.NewProgram()
	compiler.LoadJdkStub(program)
	program.AddProjectFile("file:///I.java", "interface I {}")
	program.AddProjectFile("file:///A.java", "class A implements I {}")
	program.AddProjectFile("file:///B.java", "class B extends A {}")

	i := program.GetGlobalIndex().GetType("I")
	a := program.GetGlobalIndex().GetType("A")
	index := GetSubtypeIndex(program)

	if got := names(index.DirectSubtypesOf(i)); len(got) != 1 || got[0] != "A" {
		t.Errorf("direct subtypes of I = %v, want [A]", got)
	}
	all := names(index.AllSubtypesOf(i))
	sort.Strings(all)
	if len(all) != 2 || all[0] != "A" || all[1] != "B" {
		t.Errorf("all subtypes of I = %v, want [A B]", all)
	}
	if got := names(index.DirectSubtypesOf(a)); len(got) != 1 || got[0] != "B" {
		t.Errorf("direct subtypes of A = %v, want [B]", got)
	}

	program.AddProjectFile("file:///C.java", "class C implements I {}")
	index = GetSubtypeIndex(program)
	if got := index.AllSubtypesOf(program.GetGlobalIndex().GetType("I")); len(got) != 3 {
		t.Errorf("all subtypes of I after change = %d, want 3", len(got))
	}
}
